clearwater-infrastructure/usr/share/clearwater/infrastructure/scripts/logging.bash

```diff
@@ -0,0 +1,174 @@
+# @file logging.bash
+#
+# Project Clearwater - IMS in the Cloud
+# Copyright (C) 2016  Metaswitch Networks Ltd
+#
+# This program is free software: you can redistribute it and/or modify it
+# under the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or (at your
+# option) any later version, along with the "Special Exception" for use of
+# the program along with SSL, set forth below. This program is distributed
+# in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+# without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+# A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details. You should have received a copy of the GNU General Public
+# License along with this program.  If not, see
+# <http://www.gnu.org/licenses/>.
+#
+# The author can be reached by email at clearwater@metaswitch.com or by
+# post at Metaswitch Networks Ltd, 100 Church St, Enfield EN2 6BQ, UK
+#
+# Special Exception
+# Metaswitch Networks Ltd  grants you permission to copy, modify,
+# propagate, and distribute a work formed by combining OpenSSL with The
+# Software, or a work derivative of such a combination, even if such
+# copying, modification, propagation, or distribution would otherwise
+# violate the terms of the GPL. You must comply with the GPL in all
+# respects for all of the code used other than OpenSSL.
+# "OpenSSL" means OpenSSL toolkit software distributed by the OpenSSL
+# Project and licensed under the OpenSSL Licenses, or a work based on such
+# software and licensed under the OpenSSL Licenses.
+# "OpenSSL Licenses" means the OpenSSL License and Original SSLeay License
+# under which the OpenSSL Project distributes the OpenSSL toolkit software,
+# as those licenses appear in the file LICENSE-OPENSSL.
+
+# This module is designed to provide a common logging facility which may be
+# used by all Bash scripts in Clearwater. It is intended to be sourced by other
+# scripts, passing in the following arguments:
+#   $1 - The name of the log file to create.
+#   $2 - The logging level. Defaults to WARNING
+#   $3 - The total number of log files to keep. Defaults to 1.
+
+function _logging_init()
+{
+  local filename="$1"
+  local log_level="$2"
+  local num_files="$3"
+
+  if ! [[ "$num_files" =~ ^[0-9]+$ ]]
+  then
+    echo "ERROR: Number of log backups is not a valid integer:" $num_files
+    return 1
+  fi
+
+  if (( "$num_files" > 1 ))
+  then
+    # Create a temporary logrotate config file to rotate old log files.
+    local full_filename=$(readlink -f "$filename")
+    local config_filename=$full_filename.logging.cfg
+    cat <<EOF > $config_filename
+$full_filename {
+  rotate $num_files
+  nocompress
+  missingok
+}
+EOF
+
+    # The default state file isn't writable by all users, so create a temp one
+    # here and delete it when we're done.
+    logrotate -f $config_filename -s $full_filename.logging.state
```

Comment:
I *think* you could avoid creating a temporary config file if you wanted. The bash builtin `<( command )` will run `command` in a subshell, and make its output accessible as a file. So for example `cat <( echo "Hello world" )` prints `Hello world`. In this case you could do:

```bash
logrotate -f <( cat <<EOF
$full_filename {
  rotate $num_files
  nocompress
  missingok
}
EOF )
```

However, I'm not sure this is useful given that you also need the state file.